Any API: Fix existing examples,  add new example on fetching data from json array, correct job ids (#724)
smartcontractkit/documentation: docs/Using Any API/array-response.md

- [Make an Existing Job Request](#make-an-existing-job-request)

## Array Example

This example shows how to:

- Call an API that returns a JSON array.
- Fetch a specific information from the response.

[Coingecko _GET /coins/markets/_ API](https://www.coingecko.com/en/api/documentation) returns a list of coins and their market data such as price, market cap, and volume. To check the response, you can directly paste the following URL in your browser `https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=100&page=1&sparkline=false` or run this command in your terminal:

```curl
curl -X 'GET' \
  'https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&order=market_cap_desc&per_page=100&page=1&sparkline=false' \
  -H 'accept: application/json'
```

The response should be similar to the following:

```json
[
  {
    "id": "bitcoin",
    "symbol": "btc",
    "name": "Bitcoin",
    "image": "https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579",
    "current_price": 42097,
    "market_cap": 802478449872,
    ...
  },
  {
    ...
  }
]
```

> Fetch the _id_ of the first element. To consume an API, your contract must inherit from [ChainlinkClient](https://github.com/smartcontractkit/chainlink/blob/master/contracts/src/v0.8/ChainlinkClient.sol). This contract exposes a struct named `Chainlink.Request`, which your contract can use to build the API request. The request must include the following parameters:

Reviewer: ```suggestion
Fetch the _id_ of the first element. To consume an API, your contract must import [ChainlinkClient.sol](https://github.com/smartcontractkit/chainlink/blob/master/contracts/src/v0.8/ChainlinkClient.sol). This contract exposes a struct named `Chainlink.Request`, which your contract can use to build the API request. The request must include the following parameters:
```

1. Run the `requestFirstId` function. This builds the `Chainlink.Request` using the correct parameters. The `req.add("path", "0,id")` request parameter tells the oracle node to fetch the id at index 0 of the array returned by the GET request. It uses [JSONPath expression](https://jsonpath.com/) with comma(,) delimited string for nested objects, for example: `'0,id'`.

1. After few seconds, call the `id` function. You should get a non-empty response: _bitcoin_

## Setting the LINK token address, Oracle, and JobId

The [`setChainlinkToken`](/docs/chainlink-framework/#setchainlinktoken) function sets the LINK token address for the [network](/docs/link-token-contracts/) you are deploying to. The [`setChainlinkOracle`](/docs/chainlink-framework/#setchainlinkoracle) function sets a specific Chainlink oracle that a contract makes an API call from. The `jobId` refers to a specific job for that node to run.

Each job is unique and returns different types of data. For example, a job that returns a `bytes32` variable from an API would have a different `jobId` than a job that retrieved the same data, but in the form of a `uint256` variable.

> [market.link](https://market.link/) provides a searchable catalogue of Oracles, Jobs and their subsequent return types.

Reviewer: ```suggestion
[market.link](https://market.link/) provides a searchable catalog of oracles, jobs, and their subsequent return types.
```

> [market.link](https://market.link/) provides a searchable catalogue of Oracles, Jobs and their subsequent return types.

Reviewer: ```suggestion
[market.link](https://market.link/) provides a searchable catalog of oracles, jobs, and their subsequent return types.
```
Americans don't know how to spell "catalogue" 🤣, so let's just use "catalog" consistently across the documentation. Also Oxford comma.

Author: hahah sure

    <a href="/docs/conceptual-overview/#what-is-remix" >What is Remix?</a>
</div>

To use this contract:

1. Open the [contract in Remix](https://remix.ethereum.org/#url=https://docs.chain.link/samples/APIRequests/FetchFromArray.sol).

1. Compile and deploy the contract using the Injected Web3 environment. The contract includes all the configuration variables for _Kovan_ testnet , make sure you are connected to _Kovan_. Note that in the _constructor_ , we setup:

   - The Chainlink Token address for _Kovan_ by calling [`setChainlinkToken`](/docs/chainlink-framework/#setchainlinktoken) function.
   - The Oracle contract address for _Kovan_ by calling [`setChainlinkOracle`](/docs/chainlink-framework/#setchainlinkoracle) function.
   - The `jobId`: specific job for the oracle node to run. Note that in this case, the _id_ is a _string_ data type. Hence, we need to call a job which calls an API and returns a _string_.The job spec for the Chainlink node in this example can be [found here](/docs/direct-request-get-string/).

1. Fund your contract with 0.1 LINK as shown [here](/docs/fund-your-contract/).

1. Call `id` function to confirm `id` is empty.

> 1. Run `requestFirstId` function. This builds the `Chainlink.Request` using the right parameters. Note the syntax that will be used by the oracle node to parse the _path_: `req.add("path", "0,id")` ("Fetch the id at index 0 of the array").

Reviewer: ```suggestion
1. Run `requestFirstId` function. This builds the `Chainlink.Request` using the correct parameters. The `req.add("path", "0,id")` request parameter tells the oracle node to fetch the id at index 0 of the array returned by the GET request.
```

> 1. After few seconds, call `id` function. You should get a non-empty response: _bitcoin_ at the time of writing.

Reviewer: ```suggestion
1. After few seconds, call the `id` function. You should get a non-empty response: _bitcoin_
```

## Choosing Link token address, Oracle and JobId

> [`setChainlinkToken`](/docs/chainlink-framework/#setchainlinktoken) function allows to set the Link token address for the [network](/docs/link-token-contracts/) you are deploying to. [`setChainlinkOracle`](/docs/chainlink-framework/#setchainlinkoracle) function allows to set a specific Chainlink oracle that a contract makes an API call from, and `jobId` refers to a specific job for that node to run. Each job is unique and returns different types of data.
> 
> For example, a job that returns a `bytes32` variable from an API would have a different `jobId` than a job that retrieved the same data, but in the form of a `uint256` variable.

Reviewer: ```suggestion
The [`setChainlinkToken`](/docs/chainlink-framework/#setchainlinktoken) function sets the LINK token address for the [network](/docs/link-token-contracts/) you are deploying to. The [`setChainlinkOracle`](/docs/chainlink-framework/#setchainlinkoracle) function sets a specific Chainlink oracle that a contract makes an API call from. The `jobId` refers to a specific job for that node to run.

Each job is unique and returns different types of data. For example, a job that returns a `bytes32` variable from an API would have a different `jobId` than a job that retrieved the same data, but in the form of a `uint256` variable.
```

[market.link](https://market.link/) provides a searchable catalogue of Oracles, Jobs and their subsequent return types.

## Make an Existing Job Request

> If your contract is calling a public API endpoint, an Oracle job may already exist for it. If so, it could mean you do not need to add the URL, or other adapter parameters into the request, since the job already configured to return the desired data. This makes your smart contract code more succinct. To see an example of a contract using an existing job, see [Make an Existing Job Request](../existing-job-request/).

Reviewer: ```suggestion
If your contract is calling a public API endpoint, an Oracle job might already exist for it. It could mean that you do not need to add the URL or other adapter parameters into the request because the job is already configured to return the desired data. This makes your smart contract code more succinct. To see an example of a contract using an existing job, see the [Make an Existing Job Request](../existing-job-request/) guide.
```

> For more information about the functions in `ChainlinkClient`, visit [ChainlinkClient API Reference](../chainlink-framework/).

Reviewer: ```suggestion
For more information about the functions in `ChainlinkClient`, visit the [ChainlinkClient API Reference](../chainlink-framework/).
```
